facp.md

```diff
@@ -0,0 +1,224 @@
+  uint8_t GPE0_BLK_LEN;
+  uint8_t GPE1_BLK_LEN;
+  uint8_t GPE1_BASE;
+  uint8_t CST_CNT;
+  uint16_t P_LVL2_LAT;
+  uint16_t P_LVL3_LAT;
+  uint16_t FLUSH_SIZE;
+  uint16_t FLUSH_STRIDE;
+  uint8_t duty_offset;
+  uint8_t duty_width;
+  uint8_t day_alrn;
+  uint8_t mon_alrm;
+  uint8_t century;
+  uint16_t iapc_boot_arch;
+  uint8_t reservedi;
+  uint32_t flags;
+  addr_type_t reset_reg;
+  uint8_t reset_value;
+  uint16_t arm_boot_arch;
+  uint8_t fadt_minor_version;
+  uint64_t x_firmware_ctrl;
+  uint64_t x_dsdt;
+  addr_type_t X_PM1a_EVT_BLK;
+  addr_type_t X_PM1b_EVT_BLK;
+  addr_type_t X_PM1a_CNT_BLK;
+  addr_type_t X_PM1b_CNT_BLK;
+  addr_type_t X_PM2_CNT_BLK;
+  addr_type_t X_PM_TMR_BLK;
+  addr_type_t X_GPE0_BLK;
+  addr_type_t X_GPE1_BLK;
+  addr_type_t SLEEP_CONTROL_REG;
+  addr_type_t SLEEP_STATUS_REG;
+  uint64_t hv_vendor_identity;
+  //More going to be added
+}
+```
+.
+
+It contains the header, the DSDT pointer, and blocks.
+The blocks will be used for shutdown.
+
+## Shutdown via FACP
+The ACPI shutdown is known as easy as several couple lines, however, enablement and initialization is neccessary for the shutdown code.
+
+### I. Enabling the ACPI
+
+Enabling the ACPI is done with the following steps:
+
+1. Check if ACPI is enabled
+2. Check if ACPI is enabled
+3. If not-->enable ACPI, If yes-->stop because it is already
+4. Check if ACPI is enabled
+5. If yes-->Enable ACPI, If no-->send error
+
+The first check, step 1, could be done by comparing inw PM1a CNT Block &(AND operator) SCI_EN and 0
+
+Code: `inw((uint16_t) PM1a_CNT_BLK) & (uint16_t) SCI_EN) == 0`
+
+The next check can be done by checking if smi cmd and acpi enable is 0
+
+If yes, the ACPI is already enabled for you, if not, it isn't.
+
+Code again: `smi_cmd != 0 && acpi_enable != 0`
+
+Now it's time to enable the ACPI with a single line of code:
+
+`outb((uint16_t) smi_cmd, acpi_enable);`
+
+It outbytes data acpi enable to smi cmd port
+
+and finally we need to check if it enabled the ACPI. It can be done by 3 steps:
+
+1. For the PM1a CNT Block: Could it work?
+```c
+int i;
+for(i = 0; i < 300; i++){
+	if(inw((uint16_t) PM1a_CNT_BLK & (uint16_t) SCI_EN) == 1){
+		break;
+	}
+		// If else, you can sleep a bit and continue again!!
+}
+```
+2. Do the same thing for the pm2b cnt block.
+3. Give error if not, give success message if is.
+
+### II. ACPI initialization
+
+You must have found the FACP and have the header.
+
+S5 Address can be found by several lines of code basically doing a while
+
+But you need to first add 36 from the DSDT as the address, get the dsdt length as the following code, and finally do a while until DSDT Length -- is greater than 0
+The code below performs a check at every byte, looking for a sequence of 4 characters, you should compare the first 4 characters via memcmp like functions and when you find it, you'll break and continue to the next step. Don't forget to add 1 every time to the address or you will not get the loop properly.
```

Comment:
is it looking for _S5_? If yes you can specify what it is looking for, like the code below is searching for the string "_S5_"